File: examples/python/employee.py
#!/usr/bin/env python3

from dataclasses import dataclass
from enum import Enum


class EmployeeRole(str, Enum):
    """Employee role enumeration."""

    MANAGER = "Manager"
    DEVELOPER = "Developer"


@dataclass
class EmployeeProps:
    """Properties for an Employee."""

    given_name: str
    surname: str
    salary: float
    role: EmployeeRole


class Employee:
    """Class representing an employee."""

    def __init__(
        self,
        *,
        given_name: str = "Unknown",
        surname: str = "Unknown",
        salary: float = 0.0,
        role: EmployeeRole = EmployeeRole.DEVELOPER,
    ) -> None:
        """
        Initialize an Employee.

        Args:
            given_name: The employee's given name
            surname: The employee's surname
            salary: The employee's salary
            role: The employee's role
        """
        self._given_name: str = given_name
        self._surname: str = surname
        self._salary: float = salary
        self._role: EmployeeRole = role

    @property
    def given_name(self) -> str:
        """Get the employee's given name."""
        return self._given_name

    @property
    def surname(self) -> str:
        """Get the employee's surname."""
        return self._surname

    @property
    def salary(self) -> float:
        """Get the employee's salary."""
        return self._salary

    @property
    def role(self) -> EmployeeRole:
        """Get the employee's role."""
        return self._role

    def get_full_name(self) -> str:
        """Get the employee's full name."""
        return f"{self._given_name} {self._surname}"

    def __str__(self) -> str:
        """String representation of the employee."""
        formatted_salary = f"${self._salary:,.2f}"
        return f"{self._role.value}: {self.get_full_name()} - {formatted_salary}"

    @classmethod
    def create(cls, **kwargs) -> "Employee":
        """Factory method to create a new Employee instance."""
        return cls(**kwargs)

File: examples/javascript_mjs/src/employee.mjs
import { formatCurrency } from "./utils.mjs";

/**
 * Class representing an employee implementation.
 */
class EmployeeImpl {
  /**
   * Private constructor for Employee class.
   * @param {Object} [obj] - Optional employee properties object
   * @param {string} [obj.givenName] - The given name of the employee
   * @param {string} [obj.surname] - The surname of the employee
   * @param {number} [obj.salary] - The salary of the employee
   */
  constructor(obj) {
    this.givenName = obj?.givenName ?? "Unknown";
    this.surname = obj?.surname ?? "Unknown";
    this.salary = obj?.salary ?? 0.0;
  }

  /**
   * String representation of the employee.
   * @returns {string} - A string representation of the employee
   */
  toString() {
    const formattedSalary = formatCurrency(this.salary);
    return `Name: ${this.getFullName()}, Salary: ${formattedSalary}`;
  }

  /**
   * Get the full name of the employee.
   * @returns {string} - The full name of the employee
   */
  getFullName() {
    return `${this.givenName} ${this.surname}`;
  }
}

/**
 * Builder for creating Employee instances.
 */
EmployeeImpl.Builder = class {
  constructor() {
    this.givenName = "Unknown";
    this.surname = "Unknown";
    this.salary = 0.0;
  }

  /**
   * Set the given name.
   * @param {string} givenName - The given name to set
   * @returns {EmployeeImpl.Builder} - The builder instance for chaining
   */
  withGivenName(givenName) {
    this.givenName = givenName;
    return this;
  }

  /**
   * Set the surname.
   * @param {string} surname - The surname to set
   * @returns {EmployeeImpl.Builder} - The builder instance for chaining
   */
  withSurname(surname) {
    this.surname = surname;
    return this;
  }

  /**
   * Set the salary.
   * @param {number} salary - The salary to set
   * @returns {EmployeeImpl.Builder} - The builder instance for chaining
   */
  withSalary(salary) {
    this.salary = salary;
    return this;
  }

  /**
   * Build and return a new Employee instance.
   * @returns {EmployeeImpl} - A new Employee instance
   */
  build() {
    return new EmployeeImpl({
      givenName: this.givenName,
      surname: this.surname,
      salary: this.salary,
    });
  }
};

/**
 * Factory method to create a new Employee.
 * @param {Object} [props] - Optional employee properties
 * @returns {EmployeeImpl} - A new Employee instance
 */
EmployeeImpl.NewEmployee = (props) => {
  return new EmployeeImpl(props);
};

export { EmployeeImpl };
